Add tests for FastApiAppPacker class

OlawumiSalaam opened the pull request:
- [x]  Have you followed the guidelines in our [Contribution Guide](https://github.com/ersilia-os/ersilia/blob/master/CONTRIBUTING.md)
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

**Description**
This pull request introduces comprehensive test cases for the `FastApiAppPacker` class in the `packer.py` script. The primary purpose of this class is to automate the bundling process for FastAPI applications, handling everything from directory creation to metadata management, API schema generation, and final packaging for deployment.

Key Features of `FastApiAppPacker`:
1. **Initialization**
Method: `__init__`

- Initializes the `FastApiAppPacker` class by setting up paths (`dest_dir, bundles_repo_path`), the model ID, directories for the bundle, and determining whether to use a `Dockerfile or YAML` as the installer.
- This method prepares the environment for the bundling process.

2. **Signal Handling**
Method: `_signal_handler`

- Implements a signal handler to ensure the installation process terminates gracefully if interrupted.
- Ensures resources are cleaned up and temporary files are removed properly.

3. **Metadata Handling**
Method: _get_model_id

- Extracts the model ID from either a `metadata.json or metadata.yml` file within the model directory.
- Throws an error if the metadata file is missing or improperly formatted.

4. **Bundle Structure**
Method: `_create_bundle_structure`

- Creates the necessary directory structure for the model bundle (e.g., copying model files, schemas, and app files).
- Ensures all essential files are placed in the correct locations within the bundle directory.
Method: `_get_input_schema`
- Retrieves the input schema for the model by copying the appropriate template file(s) to the bundle.

5. **Install Script Writing & Execution**
Method: `_write_install_script`
- Checks for existing install files (e.g., `install.yml or Dockerfile`) and writes the installation script for package dependencies.
Method: `_install_`packages
- Executes the installation process using a bash script generated in `_write_install_script`.
6. **API Handling**
Method:` _get_api_names_from_sh`
- Parses API names from a shell script (e.g., run.sh) to identify commands used for API endpoints.
Method: `_create_app_files`
- Generates app-related files, including app/main.py, required for the FastAPI app.
- Ensures that artifacts like API schemas and commands are properly included in the app files.
Method: `_edit_post_commands_app`
- Modifies app/main.py to include necessary post-processing commands for the APIs.
Method:` _write_api_schema`
- Generates an  schema for the APIs based on the metadata file.
Note: This method is currently incomplete, as its implementation ends abruptly in the code.

**The following Test Cases were Added:**
The following tests were implemented to ensure correctness and robustness:
Correct Initialization: Verifies that the **FastApiAppPacker** object initializes attributes like `dest_dir, bundles_repo_path`, and `install_writer `correctly using a temporary directory.
`test_get_model_id`:
Tests this method with:

- Valid JSON file.
- Valid YAML file.
- Invalid file format (asserts exception handling).

`test_get_favicon`: Mocks GET requests to verify the correct retrieval and storage of favicon.ico.
`test_create_bundle_structure`: Checks whether model files are copied to the correct directory.
`test_get_input_schema`: Verifies the correct template files are copied into the model bundle.
`test_get_api_names_from_sh`: Confirms that API names are correctly parsed from run.sh (e.g., returning [run]).
`test_create_app_files`: Ensures the app files, including artifacts, are properly created in the bundle.
`test_edit_post_commands_app`: Verifies proper parsing and code generation in app/main.py.
`test_write_api_schema`: Checks whether the API schema is generated correctly based on the metadata.
Improvements were made to the packer.py script to align with the new test cases.

**Testing locally**
These tests were implemented using the pytest and temporary directories (The pytest tmp directory [fixture](https://pytest.org/en/7.4.x/how-to/tmp_path.html) can be utilized to perform these file operations.) to simulate file operations without affecting the repository.
All the test passed. 

Closes #30

DhanshreeA commented:
Hey @OlawumiSalaam I have rolled back the signal handling functionality in the FastApiPacker class for now. You would need to update your PR. Thanks!

DhanshreeA commented on src/ersilia_pack/packer.py:
```diff
@@ -15,6 +15,7 @@ class FastApiAppPacker(object):
     def __init__(self, repo_path, bundles_repo_path, conda_env_name=None):
         self.dest_dir = repo_path
         self.bundles_repo_path = bundles_repo_path
+        self.src_dir = os.path.join(os.path.dirname(__file__), 'src', 'ersilia_pack')
```
I don't fully understand why we need this variable in favor of `root`?

DhanshreeA commented on src/ersilia_pack/parsers/metadata_yml2json_converter.py:
```diff
@@ -22,28 +22,32 @@ def _tostr(self, value):
-        data["Tags"] = self._tolist(self.data["Tags"])
+        
+        # Use .get() to handle missing keys
+        data["Identifier"] = self._tostr(self.data.get("Identifier", "default_id"))
```
We don't want any default values, as. much as it is a good practice to have them. If the model's metadata doesn't contain any of these values, it's best to actually have that result in an error.

DhanshreeA replied:
Wherever you see an `if` condition, for example the `if "Status" in self.data`, that is intentional - because that is not a field we always expect the users to have filled out in the metadata.

DhanshreeA commented on src/ersilia_pack/parsers/metadata_yml2json_converter.py:
```diff
@@ -22,28 +22,32 @@ def _tostr(self, value):
         with open(self.json_file, 'w') as f:
             json.dump(data, f, indent=4)
+
+        return data
```
We don't really need the function to return anything.

OlawumiSalaam replied:
@DhanshreeA it was returning `None` and was throwing this error
[test_error_log2.txt](https://github.com/user-attachments/files/18201391/test_error_log2.txt)

DhanshreeA commented on tests/test_FastApiPacker.py:
```diff
@@ -0,0 +1,258 @@
+# 1. Test correct initialization
+def test_fastapi_packer_initialization(temp_model_directory, bundles_repo_path):
+    metadata_file = temp_model_directory / "metadata.json"
+    metadata_file.write_text(json.dumps({"Identifier": "test_model_id"}))
```
Can we mimic the side-effects to resemble more real world scenario? Like `eos3b5e` or something as the identifier.

DhanshreeA commented on tests/test_FastApiPacker.py:
```diff
@@ -0,0 +1,258 @@
+
+# Run tests
+if __name__ == "__main__":
+    pytest.main()
```
We don't really need this to run the tests as a script. Pytest's test discovery will take care of finding these tests within the repo.

DhanshreeA commented:
Hey @OlawumiSalaam could we convert the individual test functions into a `TestFastApiPacker` class?

OlawumiSalaam commented:
@DhanshreeA 
 **Changes Made:**

I fetched the latest version of the repository, which reflected that the signal handling functionality in the `FastApiPacker` class had been rolled back. After updating the tests, one of the tests failed, and the error log showed the following:
`FAILED tests/test_FastApiPacker.py::test_get_model_id_with_yaml - KeyError: 'Title'`
 a detailed error log [test_error_log.txt](https://github.com/user-attachments/files/18200947/test_error_log.txt)
Upon inspecting the error, I realized that the `MetadataYml2JsonConverter` class was expecting the `metadata.yml` file to contain a "Title" field, which was missing. To resolve this, I added the "Title" key to the `metadata.yml` file.
However, the `KeyError` persisted. I discovered that the `MetadataYml2JsonConverter` class expects several other keys in the `metadata.yml` file, and some of them were also missing.

To implement a more robust solution, I modified the convert method in the `MetadataYml2JsonConverter` class to handle missing keys gracefully. Specifically, I used the .get() method to handle situations where a key is missing in the YAML file. This change ensures that a `KeyError` is not raised if any expected key is absent.

OlawumiSalaam commented:
@DhanshreeA I have taken all your feedback and guidelines into consideration 
I utilize a model eos3b5e to populate a test metadata.yml file and converted the individual test functions into a TestFastApiPacker class. But I am getting this error I have been stuck with since yesterday. Here is the error file
[test_FastApiPacker_output2.txt](https://github.com/user-attachments/files/18209690/test_FastApiPacker_output2.txt)
Please what do you advice. Should push the implementation to my PR so you can inspect? Thank you.

DhanshreeA commented:
Hey @OlawumiSalaam please update the PR. Particularly, please undo the changes within the `metadata_yml2json_converter.py` module.

OlawumiSalaam commented:
@DhanshreeA 
This is my initial error 
[test_FastApiPacker_output4.txt](https://github.com/user-attachments/files/18226751/test_FastApiPacker_output4.txt)
and i updated with this line ` os.path.join(tmp_dir, "metadata.yml")` and still got this error
[test_FastApiPacker_output5.txt](https://github.com/user-attachments/files/18226770/test_FastApiPacker_output5.txt)
Additionally, this model `eos9gg2` was not found. could it be a typo with the identifier?

OlawumiSalaam commented:
@DhanshreeA I have implemented the changes and updated the PR. Thank you for your help
